ang/civicase/shared/directives/on-contact-tab-change.directive.js

```diff
@@ -0,0 +1,47 @@
+(function ($, angular) {
+  var module = angular.module('civicase');
+
+  module.directive('civicaseOnContactTabChange', function (parseUrlParameters) {
+    return {
+      restrict: 'A',
+      link: onContactTabChangeLink,
+      scope: {
+        onContactTabChange: '&civicaseOnContactTabChange'
+      }
+    };
+
+    /**
+     * Listens for contact tab change and triggers the handler
+     * passed down to this directive. It provides the URL and URL
+     * parameters (as objects) for the new tab.
+     *
+     * @param {object} $scope a reference to the directive's scope.
+     */
+    function onContactTabChangeLink ($scope) {
+      var contactTabs = $('.page-civicrm-contact #mainTabContainer');
+
+      (function init () {
+        contactTabs.on('tabsactivate', listenToContactTabChange);
+      })();
+
+      /**
+       * Handles the contact tab change.
+       *
+       * @param {document#event:tabsactivate} event the tab change event reference.
+       * @param {object} uiTab a UI Tab event object which contains a reference to the new
+       *   and previous tabs.
+       */
+      function listenToContactTabChange (event, uiTab) {
+        var tabUrl = $(uiTab.newTab).find('a').attr('href');
+        var urlParams = parseUrlParameters(tabUrl);
+
+        $scope.$apply(function () {
+          $scope.onContactTabChange({
+            $tabUrl: tabUrl,
```

Comment:
Why do we have object keys starting with `$`? Is there a particular reason?

Reply:
Adding a `$` to variables that are going to be injected to the scope is a standard. This is done to avoid replacing existing values. You can see an example from the AngularjsUI project:

https://angular-ui.github.io/ui-select/demo-append-to-body.html

If you view the source and jump to line 98/99 you'll see this:

```html
<ui-select-choices repeat="person in ctrl.people \| propsFilter: {name: $select.search, age: $select.search}">
  <div ng-bind-html="person.name \| highlight: $select.search"></div>
```

In this case the variable `$select` is injected into the scope from the directive and it contains the user selected and search values.

In our case you can see that we also use this value in the HTML:

https://github.com/compucorp/uk.co.compucorp.civicase/pull/359/files#diff-ce05111aafacf8a33a6add651784b61fR4

```html
<div
  id="bootstrap-theme"
  class="clearfix civicase__container"
  civicase-on-contact-tab-change="handleContactTabChange($tabUrlParams)"
>
```